ButtonGroup & ButtonToggle: add options, use buttons

mdodgelooker opened the pull request:
### :sparkles: Changes

**This ended up being pretty much a complete rewrite.**

- Using hidden inputs to dictate the behavior of `ButtonGroup` and `ButtonToggle` was problematic for both accessibility (unable to focus) and CI in the core product
- I replaced the `<label><input>text</label>` structure with `<button>text</button>`
- In doing so, the uncontrolled version no longer worked, nor could I see a reason for supporting it (with inputs, you could use them in an old-school form) so now the components must be controlled
- Added `options` as an alternative to `ButtonItem` children
- Styling polish (especially for the wrapping scenario)

### :white_check_mark: Requirements

- [x] Includes test coverage for all changes
- [x] Build and tests are passing
- [x] Update documentation
- [x] Updated CHANGELOG
- [x] Checked for i18n impacts
- [x] Checked for a11y impacts
- [ ] PR is ideally < ~400LOC

### :camera: Screenshots
![button-set](https://user-images.githubusercontent.com/53451193/83687908-5bff5800-a5a1-11ea-8668-0d23c3d7b340.gif)

mdodgelooker commented on packages/components/src/Button/ButtonGroup.tsx:
```diff
@@ -25,63 +25,55 @@
-      onChange,
-      value,
-      ...props
-    }: Omit<ButtonGroupOrToggleBaseProps<string[]>, 'nullable'>,
```
Moved `nullable` (interface and logic) to where it belongs in `ButtonToggle.tsx` instead of the underlying component `ButtonSet.tsx`.

mdodgelooker commented on packages/components/src/Button/ButtonGroup.tsx:
```diff
@@ -25,63 +25,55 @@
   ) => {
-    const name = useID(props.name)
-
-    const isControlled = useControlWarn({
-      controllingProps: ['onChange', 'value'],
-      isControlledCheck: () => onChange !== undefined,
-      name: 'ButtonGroup',
-    })
```
No more uncontrolled version, since there are no longer `input`s involved.

mdodgelooker commented on packages/components/src/Button/ButtonGroup.tsx:
```diff
@@ -25,63 +25,55 @@
+export const ButtonGroup = styled(ButtonGroupLayout)`
   ${ButtonItem} {
-    border-radius: ${({ theme }) => theme.radii.medium};
-    height: 36px;
```
Moved `height` to `ButtonItem.tsx` since the value is common to both `ButtonGroup` and `ButtonToggle`.

mdodgelooker commented on packages/components/src/Button/ButtonGroup.tsx:
```diff
@@ -25,63 +25,55 @@
 )
 
-export const ButtonGroup = styled(ButtonGroupLayout)`
-  margin: -${({ theme }) => theme.space.xxxsmall};
```
Moved this ugly negative margin behavior inside `.wrapping` below. Due to some nagging issues in the styles of the wrapping vs non-wrapping `ButtonToggle`, I decided to add the `wrapping` className via a callback ref in `ButtonSet.tsx`. Of course I would rather not have to use that, but at least it allows for less convoluted css.

mdodgelooker commented on packages/components/src/Button/ButtonItem.tsx:
```diff
@@ -34,61 +34,119 @@ import {
+
+    function handleClick(e: MouseEvent<HTMLButtonElement>) {
+      onClick && onClick(e)
+      if (!e.defaultPrevented) {
```
Allows the parent handler to be cancelled.

mdodgelooker commented on packages/components/src/Button/ButtonSet.tsx:
```diff
@@ -25,117 +25,120 @@
  */
 
 import React, {
-  cloneElement,
```
My original usage of `cloneElement` here was overly convoluted, so I changed to using `Context` instead.

mdodgelooker commented on packages/components/src/Button/ButtonSet.tsx:
```diff
@@ -25,117 +25,120 @@
 interface ButtonSetProps<ValueType extends string | string[] = string[]>
   extends SimpleLayoutProps,
-    Omit<CompatibleHTMLProps<HTMLDivElement>, 'value' | 'onChange'> {
+    Omit<CompatibleHTMLProps<HTMLDivElement>, 'value' | 'defaultValue'> {
```
Not sure why divs support `defaultValue` but this component definitely does not, so I wanted to make that clear.

lukelooker replied:
Weird :p

mdodgelooker commented on packages/components/src/Button/ButtonSet.tsx:
```diff
@@ -25,117 +25,120 @@
    */
-  nullable?: boolean
+  value?: ValueType
+  onItemClick?: (e: MouseEvent<HTMLButtonElement>) => void
```
This is only used internally and is omitted from the external interface below.

mdodgelooker commented on packages/components/src/Button/ButtonSet.tsx:
```diff
@@ -25,117 +25,120 @@
 export type ButtonSetType<
-  T extends string | string[] = string[]
-> = ForwardRefExoticComponent<ButtonSetProps<T> & { ref: Ref<HTMLDivElement> }>
+  TValue extends string | string[] = string[]
```
Just keeping the value type arg naming consistent.

mdodgelooker commented on packages/components/src/Button/ButtonSet.tsx:
```diff
@@ -25,117 +25,120 @@
+          setIsWrapping(false)
+        }
+      }
+    }, [])
```
Couldn't get the css water tight for both wrapping and non-wrapping `ButtonToggle` so I opted for this solution to separate some of the styles.

mdodgelooker commented on packages/components/src/Button/ButtonSet.tsx:
```diff
@@ -25,117 +25,120 @@
-      </div>
+      <ButtonSetContext.Provider value={context}>
+        <div
+          role="group"
```
This is needed for accessibility now that the items are `button`s instead of `input`s that share a common `name` attribute.

mdodgelooker commented on packages/components/src/Button/ButtonGroup.test.tsx:
```diff
@@ -25,91 +25,84 @@
+    }
+    renderWithTheme(<TestComponent />)
+
+    expect(screen.getByRole('group')).toMatchSnapshot()
```
As much as I don't love snapshots, these components are so heavy with CSS. Thoughts?

lukelooker replied:
I think it's a logical place to use one - nice to get the "signal" that something changed. When we have visual snapshots this (probably most snapshots) could likely be replaced with those.

mdodgelooker commented on packages/components/src/Button/ButtonItem.tsx:
```diff
@@ -34,61 +34,114 @@ import {
-        ? `hsla(${parseToHsl(props.theme.colors.key).hue}, 25%, 97%, 0.7)`
-        : false};
+  border: none;
+  background: transparent;
```
Just resetting the default browser button border and background. `ButtonGroup` and `ButtonToggle` have pretty different border and background styles, so I moved most of them to those individual files.

mdodgelooker commented on packages/components/src/Button/ButtonItem.tsx:
```diff
@@ -34,61 +34,119 @@ import {
+
+    color: ${({ theme }) => theme.colors.key};
+    &[disabled] {
+      color: ${({ theme }) => theme.colors.keyFocus};
```
This does not seem like the right name of the color for selected & disabled, but it looks good to me... @jhardy, thoughts?
![image](https://user-images.githubusercontent.com/53451193/83687384-87357780-a5a0-11ea-8d90-f16c499e70fb.png)

lukelooker replied:
Generally `disabled` states are grey but admittedly this case is a little different than some. It's basically `:disabled + :active`

lukelooker commented on packages/components/src/Button/ButtonItem.tsx:
```diff
@@ -34,61 +34,119 @@ import {
   &:active {
-    background: ${(props) =>
-      `hsla(${parseToHsl(props.theme.colors.key).hue}, 50%, 96%, 0.9)`};
+    background: hsla(
+      ${({ theme }) => parseToHsl(theme.colors.key).hue},
+      50%,
+      96%,
+      0.9
+    );
```
Could we use `keyAccent` or `keySubtle` here instead? Might need a @jhardy consult.

jhardy replied:
+1 to using keyAccent and keySubtle here.

lukelooker commented on packages/components/src/Button/ButtonItem.tsx:
```diff
@@ -34,61 +34,119 @@ import {
+  transition: background ${({ theme }) => theme.transitions.durationQuick} ease;
+  color: ${({ theme }) => theme.colors.text3};
+
+  &[aria-pressed='false']:not([disabled]):hover {
+    background: hsla(
+      ${({ theme }) => parseToHsl(theme.colors.key).hue},
+      25%,
+      97%,
+      0.7
+    );
```
Could we use `keyAccent` or `keySubtle` here instead? Might need a @jhardy consult.

lukelooker commented on packages/components/src/Button/ButtonItem.tsx:
```diff
@@ -34,61 +34,119 @@ import {
 
-  input:checked + ${ButtonItemLabelText} {
-    color: ${(props) => props.theme.colors.key};
+  &[aria-pressed='true'] {
+    background: hsla(
+      ${({ theme }) => parseToHsl(theme.colors.key).hue},
+      100%,
+      98%,
+      1
+    );
+
+    border-color: hsla(
+      ${({ theme }) => parseToHsl(theme.colors.key).hue},
+      100%,
+      98%,
+      1
+    );
```
Few more. If we can't use those slots let's look at "naming" this behavior and moving the calculation functionality to `design-tokens/utils` - I want to consolidate and color math into that package.

lukelooker commented on packages/components/src/Button/ButtonSet.tsx:
```diff
@@ -25,117 +25,120 @@
+  children?: ReactNode
   /**
-   * Value for controlling the component
+   * Available options (do not use if using ButtonItem children)
```
Can we put a warning in if they supply both / neither? I guess there's a chance this might get rendered with no options and then updated?

mdodgelooker replied:
I have a warning down further in the file for both, but yeah I could see neither happening briefly if the options are being loaded asynchronously.

jhardy commented on packages/components/src/Button/ButtonItem.tsx:
```diff
@@ -34,61 +34,119 @@ import {
+export const buttonItemHeight = 36
 
-export const ButtonItemLabel = styled.label<ButtonItemProps>`
-  cursor: pointer;
```
Can we add `cursor: pointer;` back in? I'd like to keep the cursors consistent with our other buttons

mdodgelooker commented:
@jhardy and @lukelooker I replaced the mathy colors with `keyAccent` and `keySubtle`. There wasn't a good option for the `:active` state (the next darker, `keyFocus` is too dark) so hopefully it's ok without for now?
(I also added a few things I realized were missing in the docs page.)

mdodgelooker commented:
Oops, snapsots...
